File: beat-planner/src/TrackPlayer.cs
using System;
using System.Threading;

namespace BeatPlanner
{
	public class TrackPlayer
	{
		private Metronome metro;
		private bool shouldPlay = false;

		public int Index { get; set; }

		public readonly Track Track;

		public TrackPlayer(Track track, Metronome metro)
		{
			this.Track = track;
			this.metro = metro;
		}

		public TrackPlayer(Track track) : this(track, new Metronome())
		{

		}

		public void PlayAsync()
		{
			Thread thread = new Thread(Play);
			thread.Start();
		}

		public void Stop()
		{
			shouldPlay = false;
		}

		public void Reset()
		{
			Index = 0;
		}

		public void Play()
		{
			shouldPlay = true;
			Tuple<Beat, int>[] sequences = Track.Sequences;
			Tuple<Beat, int> seq = sequences[Index];
			Beat beat = seq.Item1;
			int reps = seq.Item2;
			int nextChange = reps;
			metro.Beat = beat;
			metro.Start();
			while (shouldPlay)
			{
				if (metro.Bars == nextChange)
				{
					if (Index == sequences.Length - 1)
						break;
					seq = sequences[++Index];
					beat = seq.Item1;
					reps = seq.Item2;
					metro.Beat = beat;
					nextChange += reps;
				}
			}
			metro.Stop();
		}
	}
}

File: app/src/main/java/dk/aschoen/beatplanner/core/TrackPlayer.java
//
// Translated by CS2J (http://www.cs2j.com): 2014-07-29 22:02:12
//

package dk.aschoen.beatplanner.core;



public class TrackPlayer
{
    private Metronome metro;
    private boolean shouldPlay = false;

    private int index;
    public int getIndex() {
        return index;
    }

    public final Track track;
    public TrackPlayer(Track track, Metronome metro) {
        this.track = track;
        this.metro = metro;
    }

    public TrackPlayer(Track track) {
        this(track, new Metronome());
    }

    public void playAsync() {
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                play();
            }
        });
        thread.start();
    }

    public void stop() {
        shouldPlay = false;
    }

    public void reset() {
        index = 0;
    }

    public void play() {
        shouldPlay = true;
        Sequence[] sequences = track.getSequences();
        Sequence seq = sequences[getIndex()];
        Beat beat = seq.getBeat();
        int reps = seq.reps;
        int nextChange = reps;
        metro.setBeat(beat);
        metro.start();
        while (shouldPlay)
        {
            if (metro.getBars() == nextChange)
            {
                if (getIndex() == sequences.length - 1)
                    break;

                index++;
                seq = sequences[getIndex()];
                beat = seq.getBeat();
                reps = seq.reps;
                metro.setBeat(beat);
                nextChange += reps;
            }

        }
        metro.stop();
    }

}


